src/valid/analyzer.rs

```diff
@@ -581,15 +581,17 @@ impl FunctionInfo {
                 let image_storage = GlobalOrArgument::from_expression(expression_arena, image)?;
                 let sampler_storage = GlobalOrArgument::from_expression(expression_arena, sampler)?;
 
-                match (image_storage, sampler_storage) {
-                    (GlobalOrArgument::Global(image), GlobalOrArgument::Global(sampler)) => {
-                        self.sampling_set.insert(SamplingKey { image, sampler });
-                    }
-                    _ => {
-                        self.sampling.insert(Sampling {
-                            image: image_storage,
-                            sampler: sampler_storage,
-                        });
+                if gather.is_none() {
+                    match (image_storage, sampler_storage) {
+                        (GlobalOrArgument::Global(image), GlobalOrArgument::Global(sampler)) => {
+                            self.sampling_set.insert(SamplingKey { image, sampler });
+                        }
+                        _ => {
+                            self.sampling.insert(Sampling {
+                                image: image_storage,
+                                sampler: sampler_storage,
+                            });
+                        }
```

Comment:
I don't see why we'd ignore gather operations here.

Reply:
Putting the gather in the sampling set lead to issues. I asked on matrix, and it was suggested that perhaps we shouldn't put it in the set. Doing that fixed my issues.

I'm not sure if it's actually the correct fix, but it seems to work for me. I don't have enough experience with naga to know what the correct fix would be, otherwise.

Reply:
I undid this change locally and tests ran fine. Will push it.